### PR TITLE
Add internal healthchecking capabilities

### PR DESCRIPTION
**Pull request by DeviaVir**
Allow spacel-agent to perform healthchecks directly on the node.

Passing on via the `manifest`:

```
"healthcheck": {
  "endpoint": "localhost/path",
  "virtual_host": "spacel-agent.com",
  "protocol": "http",
  "forwarded_protocol": "https"
}
```

@pebble/webops

### CODE REVIEW

File: src/spacel/agent/instance.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        :param timeout:  Timeout.
+        :return: Health status (boolean)
+        """
+        if not manifest.healthcheck.get('endpoint'):
+            return True
+
+        retries = 0
+        start = time.time()
+        while True:
+            if self.check_health(manifest.healthcheck.get('endpoint'),
```
Comment (thepwagner) [inline src/spacel/agent/instance.py]: Maybe `._check_health()`? Feels like private helper to `.health()`.

Maybe pass `manifest.healthcheck` as an argument, then unpack inside?
Prevents duplicated defaults (i.e. `localhost` on :31 and :46), and could de-dent for style.

File: src/test/agent/test_instance.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            'healthcheck': {
+                'endpoint': 'test',
+                'virtual_host': 'spacel',
+                'protocol': 'https'
+            }
+        })
+
+    @patch('spacel.agent.instance.InstanceManager.check_health')
+    def test_health(self, mock_check_health):
+        mock_check_health.return_value = True
```
Comment (thepwagner) [inline src/test/agent/test_instance.py]: Can't this just be `self.instance._check_health = MagicMock(return_value=True)`. Patch feels heavy

File: src/test/agent/test_instance.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        result = self.instance.health(self.manifest, 1)
+        self.assertEqual(result, True)
+
+    @patch('spacel.agent.instance.InstanceManager.check_health')
+    def test_health_fail(self, mock_check_health):
+        mock_check_health.return_value = False
+        result = self.instance.health(self.manifest, 1)
```
Comment (thepwagner) [inline src/test/agent/test_instance.py]: Nit: all these 1 second timeouts will drag on; can we tune `backoff_scale` and aim for <100ms per test?

File: src/test/agent/test_instance.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        self.assertEqual(result, False)
```
Comment (thepwagner) [inline src/test/agent/test_instance.py]: Maybe assert `mock_check_health.call_count` to verify the retry logic?

File: src/test/agent/test_instance.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        request.add_header.return_value = True
+        mock_urllib.urlopen.return_value = response
+        mock_urllib.Request.return_value = request
+        result = self.instance.check_health('localhost/healthcheck',
+                                            'spacel-agent.com', 'http')
+        self.assertEqual(result, True)
+
+    @patch('spacel.agent.instance.urllib2')
+    def test_check_health_fail(self, mock_urllib):
+        response = MagicMock()
```
Comment (thepwagner) [inline src/test/agent/test_instance.py]: Maybe a helper to de-dupe? All that changes is status_code.

File: src/test/agent/test_instance.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        mock_check_health.return_value = False
+        result = self.instance.health(self.manifest, 1)
+        self.assertEqual(result, False)
+
+    @patch('spacel.agent.instance.urllib2')
+    def test_check_health(self, mock_urllib):
+        response = MagicMock()
+        response.getcode.return_value = 200
+        request = MagicMock()
+        request.add_header.return_value = True
```
Comment (thepwagner) [inline src/test/agent/test_instance.py]: I don't get why this is needed; the return value isn't inspected?

**Comment by thepwagner** [comment]
+1, just style stuff.

**Comment by DeviaVir** [comment]
@thepwagner great reviewing! Updated the PR accordingly.

File: src/spacel/agent/instance.py [hunk 7]
```diff
@@ -43,23 +38,28 @@ def health(self, manifest, timeout=120):
             time.sleep(back_off)
 
     @staticmethod
-    def check_health(endpoint, virtual_host='localhost', protocol='http',
-                     forwarded_protocol='https'):
+    def _check_health(manifest):
```
Comment (thepwagner) [inline src/spacel/agent/instance.py]: Nit: naming makes me think this is an `AgentManifest`, but it isn't.

**Comment by thepwagner** [comment]
+1

**Pull request merged**